### DISCUSSION TITLE
Missing? features in Table/Form builder

### DISCUSSION THREAD

**Discussion by madsem** [filamentphp/filament#2319] [Ideas]
I'm using Filament Table & Form builder together with Laravel Jetstream to build an app to display sales reports for company employees.

Some things would be nice to have, that imho are currently missing (At least I think they are haha).

Example of a dashboard for the company app:
<img width="1416" alt="Screen Shot 2022-05-08 at 2 51 11 PM" src="https://user-images.githubusercontent.com/3166991/167297093-c796e757-fa89-440b-8dc5-2a5c25620c44.png">

What I would like to have:

* Be able to connect Filament Filters with methods in Livewire component that display cards/metrics. So when table updates, the cards do too.

Currently I seem to have to solve it like this (Without being able to update cards when filters change)
```php
public function getTotals()
    {
        return AccountDailyReport::query()
            ->selectRaw('SUM(clicks) AS total_visitors')
            ->selectRaw('SUM(spend) AS total_spend')
            ->selectRaw('SUM(revenue) AS total_revenue')
            ->selectRaw('SUM(conversions) AS total_conversions')
            ->selectRaw('SUM(revenue) - SUM(spend) AS total_profit')
            ->whereMonth('date', Carbon::now()->month)
            ->first();
    }


    public function render()
    {
        return view('livewire.show-dashboard', [
            'totals' => $this->getTotals(),
        ]);
    }
```

* Totals rows above/below tables. (If you have wide tables with lots of columns, it's not feasible to display totals using cards).

* Ability to give rows in tables different bg color (In my example, highlight green/red for profit/loss).

* Ability to add filters in a row on top of table (In apps where tables are often used for reporting it's easier to see what filters are currently set, less clicks also)

That's just my wishlist, but I do think a lot of use-cases would benefit from the above.
I love Filament, great job guys! 🥇

**Comment by ryangjchandler** [comment]
Table filters are stored in a `public $tableFilters` property so you could setup a watcher for that, i.e. `updatedTableFilters` and then update your cards in there?

**Comment by danharrin** [comment]
> Totals rows above/below tables. (If you have wide tables with lots of columns, it's not feasible to display totals using cards).

There's a `getContentFooter()` method that can return a view for this. It gets rendered as the last row of the table, and you can access `$columns` and `$records` inside it.

> Ability to give rows in tables different bg color (In my example, highlight green/red for profit/loss).

Highly requested feature, I'd accept a PR.

> Ability to add filters in a row on top of table (In apps where tables are often used for reporting it's easier to see what filters are currently set, less clicks also)

https://filamentphp.com/docs/2.x/tables/filters#displaying-filters-above-the-table-content

**Reply by madsem** [reply]
Oh man, thanks @danharrin. Almost everything is already possible, I should search longer haha.
Thanks so much, appreciate you guys!

**Reply by danharrin** [reply]
To be fair, they're quite new features. Regarding displaying stats above the table - if you ditch widgets and instead render the cards above the table in the same view, you won't have to pass data between Livewire components, you can just get the table query using `$this->getFilteredTableQuery()` and calculate

**Reply by madsem** [reply]
> > Ability to give rows in tables different bg color (In my example, highlight green/red for profit/loss).
> 
> Highly requested feature, I'd accept a PR.

For anyone else looking for this, I saw that it's actually possible already to do this with cells.
It requires some repetition, but you can already highlight specific cells:

```php
TextColumn::make('profit')
                ->getStateUsing(fn($record): string => $record->revenue->minus($record->spend))
                ->extraAttributes(fn($record): array => $record->is_profitable
                    ? ['class' => 'bg-green-50']
                    : ['class' => 'bg-red-50']
                ),

            TextColumn::make('roi')
                ->getStateUsing(fn($record): string => calculate_roi($record->revenue, $record->spend).'%')
                ->extraAttributes(fn($record): array => $record->is_profitable
                    ? ['class' => 'bg-green-50']
                    : ['class' => 'bg-red-50']
                ),
```

**Reply by madsem** [reply]
> To be fair, they're quite new features. Regarding displaying stats above the table - if you ditch widgets and instead render the cards above the table in the same view, you won't have to pass data between Livewire components, you can just get the table query using `$this->getFilteredTableQuery()` and calculate

I didn't realize that this is possible, so are you saying one Component class is able to use the Form & Table builder together?
Do you mind sharing a link or ssimple code example?

**Reply by madsem** [reply]
Ahhh I think I understand now what you mean. But then how could you trigger recalculation when changing filters in the UI, without doing a full refresh. That is the part I don't quite grasp tbh, even when using the same view...

**Reply by danharrin** [reply]
If the calculation happens in a getter method on the Livewire component and called from the view, then it gets refreshed every time the Livewire component rerenders.

**Reply by madsem** [reply]
> To be fair, they're quite new features. Regarding displaying stats above the table - if you ditch widgets and instead render the cards above the table in the same view, you won't have to pass data between Livewire components, you can just get the table query using `$this->getFilteredTableQuery()` and calculate

FYI: I tried this and the theory works beautifully :) But the problem is, if you get the `$this->getFilteredTableQuery()` and you already have a `groupBy`, for example to display results grouped by day... You can't also use the filtered query to get an overall total of all days.

Example:
<img width="1410" alt="Screen Shot 2022-05-11 at 11 22 06 AM" src="https://user-images.githubusercontent.com/3166991/167816199-3ada92c5-fa25-41e0-abf6-79cfa7f64d79.png">

If you utilize the already filtered query to build grouped tables, the total will then also be grouped and always show the first entry... I think what might be cool is if we could create connected cards that can get their results from a different query. Unless I'm completely forgetting about something that would allow me to retrieve two different result sets (by days, and total) in one query. (Which is entirely possible haha)

PS: Actually, I just remembered what @ryangjchandler wrote earlier and in cases where you can't re-use the filtered query, you can get the required filters one-by-one from the `$this->tableFilters` property.

Like so:
```php
$this->tableFilters['date']['date_from']
```

**Reply by danharrin** [reply]
Instead of `getFilteredTableQuery()` try...

```php
$query = ModelName::query();
$this->applyFiltersToTableQuery($query);
$this->applySearchToTableQuery($query);

// custom $query is now filtered
```

**Reply by madsem** [reply]
Oh wow, that is so much cleaner! Thanks a lot @danharrin  :)

PS: works like a charm
